source/common/network/connection_impl.cc

```diff
@@ -275,6 +275,15 @@ void ConnectionImpl::raiseEvent(ConnectionEvent event) {
     // connected events.
     callback->onEvent(event);
   }
+  // We may have pending data in the write buffer on transport handshake
+  // completion, which may also have completed in the context of onReadReady(),
+  // where no check of the write buffer is made. Provide an opportunity to flush
+  // here. If connection write is not ready, this is harmless. We should only do
+  // this if we're still open (the above callbacks may have closed).
+  if (state() == State::Open && event == ConnectionEvent::Connected &&
```

Comment:
@htuch @ggreenway I'm late here, but I think it's slightly strange that in the plaintext case we effectively recurse into onWriteReady() which already handles this case by trying to write after a connection happens. Did we consider any other variant of this change? It's not a big deal just curious.

Reply:
That is weird; I didn't catch that. No, we didn't try anything else. We should look into if there's a better way to handle this. It seems to work, but it is strange, and hard to reason about.

Reply:
It works because we flip `connecting_` status before re-entry. I agree this is harder to reason about than would be ideal. Should we just return unconditionally in raw buffer socket `onWriteReady()` following a new connection and defer to the new write check?

Reply:
@htuch agreed the code works as is. To make it clearer yes I would either return unconditionally with a comment here https://github.com/envoyproxy/envoy/blob/master/source/common/network/connection_impl.cc#L477 or another option is to factor out the 2nd part of onWriteReady() into a separate function and call that in the SSL case when handshake completes during a read callback. I think the former is simpler IMO.